### PR TITLE
Refactor the loggers

### PR DESCRIPTION
**Pull request by theofidry**
Whenever I took a look at the loggers I couldn't help but to find they could be made a bit simpler. I am not sure the overall PR is simpler, but here are the key changes:

- `FileLogger` is its own final dedicated class, no longer a base abstract class. This allows to test and pin its behaviour once and only once, not in all its inherited classes as well
- As a consequence of the above, all the previous file loggers are now plain loggers that can be unit tested (see their test differences and you will see how much simpler and enjoyable they are to test)
- A new `LoggerRegistry` which allows `SubscriberBuilder` to consume a simple logger instead of an array of loggers
- Simplify `LoggerFactory`: avoid a double looping and remove the dependence on `ResultsLoggerTypes` (the changed code just do fine without)

### CODE REVIEW

File: src/Logger/LoggerFactory.php [hunk 1]
```diff
@@ -72,121 +74,109 @@ public function __construct(
...
-        );
+        return $badge === null
+            ? null
+            : new BadgeLogger(
+                $output,
+                new ChainBuildContextResolver(new TravisCiResolver()),
+                new StrykerApiKeyResolver(),
+                new StrykerDashboardClient(
+                    new JsonClient(),
+                    new ConsoleLogger($output)
```
Comment (sanmai) [inline src/Logger/LoggerFactory.php]: I wish BadgeLogger would know these details itself. Say within a `::build()` method.

File: src/Logger/TextFileLogger.php [hunk 2]
```diff
@@ -36,27 +36,46 @@
...
 use function strlen;
 
 /**
  * @internal
  */
-final class TextFileLogger extends FileLogger
+final class TextFileLogger implements LineMutationTestingResultsLogger
 {
-    protected function getLogLines(): array
+    private $metricsCalculator;
```
Comment (sanmai) [inline src/Logger/TextFileLogger.php]: I wish these loggers would be responsible for the result collection themselves, not MetricsCalculator. Metrics are metrics, logging be logging. Also MetricsCalculator won't collect logs nobody will ask for.

This is probably for another time.

File: src/Logger/LoggerFactory.php (hunk 1 above)
Comment (theofidry) [reply src/Logger/LoggerFactory.php]: I'm still uncertain about the fate of `LoggerFactory` and attempted something in #1094. So I prefer to not touch that part in this PR

File: src/Logger/TextFileLogger.php (hunk 2 above)
Comment (theofidry) [reply src/Logger/TextFileLogger.php]: I would also like to split the metrics calculator in two. But I got to admit it kind of complicate its consumption for no good reason... Maybe we can discuss that in another issue or in #1099

**Pull request merged**